include/quic/datagram.hpp

```diff
@@ -47,7 +47,7 @@ namespace oxen::quic
     };
 
     // IO callbacks
-    using dgram_data_callback = std::function<void(dgram_interface&, bstring)>;
+    using dgram_data_callback = std::function<void(std::shared_ptr<dgram_interface>, bstring)>;
```

Comment:
We do not want to pass a shared_ptr here (or for the stream data): it will incur the overhead of instantiating an additional shared_ptr instance (which requires an atomic reference count increment on construction and decrement on destruction) on every data callback, even if the caller does not care about or need a shared ptr.  By passing it as a reference we avoid that overhead and only incur if if the caller needs it (by calling `shared_from_this()` on the reference).

The relatively infrequent calls (stream open, etc.) could be passed by shared_ptr, but I'm not sure it's worthwhile to be inconsistent (references for data callbacks, shared_ptr for other things).